# Correcting open chromatin bias in bulk and single-cell CUT&Tag data with PATTY

Precise profiling of epigenomes, including histone modifications and transcription factor binding sites, is essential for better understanding gene regulatory mechanisms. Cleavage Under Targets & Tagmentation (CUT&Tag) is an easy and low-cost epigenomic profiling method that can be performed on a low number of cells and at the single-cell level. A large number of CUT&Tag datasets have been generated in cancer samples, providing a valuable resource. 
CUT&Tag experiments use the hyperactive transposase Tn5 for tagmentation. We found that preference of Tn5 captured reads toward accessible chromatin regions can influence the distribution of CUT&Tag reads and cause open chromatin biases, further confounding the analysis of CUT&Tag data. The high sparsity of single-cell sequencing data makes the open chromatin biases more substantial than in bulk sequencing data. Here, we present a comprehensive computational method, PATTY (Propensity Analyzer for Tn5 Transposase Yielded bias), to mitigate the open chromatin bias inherent in CUT&Tag data at both bulk and single-cell levels. By integrating existing transcriptome and epigenome data using machine learning and comprehensive modeling, we demonstrate that PATTY yields more accurate and robust detection of histone modification occupancy sites for bulk CUT&Tag data than existing methods. We further design a single-cell CUT&Tag analysis framework by utilizing the model trained from bulk data and show improved cell clustering from bias-corrected single-cell CUT&Tag data compared to using raw data. This model paved the way for further development of computational tools for improving bulk and single-cell CUT&Tag data analysis.
![image](https://github.com/Tarela/PATTY/assets/3999744/bb0617b1-6779-4bd8-a94c-6f22e6362be2)

